torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -23,6 +23,57 @@ using torch::autograd::variable_list;
 static constexpr char* kNumBackwardPasses = "num_current_backward_passes";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 3 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call pre hooks of the original AccumulateGrad to modify the input grad.
+//   3. Call post hooks of the original AccumulateGrad.
+struct DistAccumulateGradCaptureHook
+    : GraphTask::ExecInfo::Capture::GradCaptureHook {
+  DistAccumulateGradCaptureHook(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
+      ContextPtr autogradContext)
+      : accumulateGrad_(std::move(accumulateGrad)),
+        autogradContext_(std::move(autogradContext)) {}
+
+  torch::Tensor operator()(const torch::Tensor& grad) override {
+    // It is possible that the grad is not defined since a separate
+    // invocation of the autograd engine on the same node might actually
+    // compute this gradient.
+    if (grad.defined()) {
+      autogradContext_->accumulateGrad(
+          accumulateGrad_->variable, grad, 1 /* num_expected_refs */);
+    }
+    // It's intended that pre/post hooks are still called even if the grad is
+    // undenfined here.
+    if (!accumulateGrad_->pre_hooks().empty()) {
```

Comment:
The pre-hooks should be called before the grad accumulation:
- Pre-hook
- Accumulation
- Post-hook.